x-pack/libbeat/reader/etw/provider_test.go

```diff
@@ -0,0 +1,229 @@
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build windows
+
+package etw
+
+import (
+	"encoding/binary"
+	"syscall"
+	"testing"
+	"unsafe"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUTF16PtrToString(t *testing.T) {
+	// Create a UTF-16 string
+	sampleText := "This is a string test!"
+	utf16Str, _ := syscall.UTF16FromString(sampleText)
+
+	// Convert to *uint16
+	ptr := &utf16Str[0]
+
+	result := UTF16PtrToString(ptr)
+	assert.Equal(t, sampleText, result, "The converted string should match the original")
+}
+
+func TestUTF16AtOffsetToString(t *testing.T) {
+	// Create a UTF-16 string
+	sampleText := "This is a string test!"
+	utf16Str, _ := syscall.UTF16FromString(sampleText)
+
+	// Convert it to uintptr (simulate as if it's part of a larger struct)
+	ptr := uintptr(unsafe.Pointer(&utf16Str[0]))
+
+	// Test the function
+	result := UTF16AtOffsetToString(ptr, 0)
+	assert.Equal(t, sampleText, result, "The converted string should match the original")
+
+	// Test with offset (skip the first character)
+	offset := unsafe.Sizeof(utf16Str[0]) // Size of one UTF-16 character
+	resultWithOffset := UTF16AtOffsetToString(ptr, offset)
+	assert.Equal(t, sampleText[1:], resultWithOffset, "The converted string with offset should skip the first character")
+}
+
+func TestGUIDFromProviderName_EmptyName(t *testing.T) {
+	guid, err := GUIDFromProviderName("")
+	assert.EqualError(t, err, "empty provider name")
+	assert.Equal(t, GUID{}, guid, "GUID should be empty for an empty provider name")
+}
+
+func TestGUIDFromProviderName_EmptyProviderList(t *testing.T) {
+	// Backup and defer restoration of the original function
+	originalFunc := EnumerateProvidersFunc
+	t.Cleanup(func() {
+		EnumerateProvidersFunc = originalFunc
+	})
+
+	// Define a mock provider name and GUID for testing.
+	mockProviderName := "NonExistentProvider"
+
+	EnumerateProvidersFunc = func(pBuffer *ProviderEnumerationInfo, pBufferSize *uint32) error {
+		// Check if the buffer size is sufficient
+		requiredSize := uint32(unsafe.Sizeof(ProviderEnumerationInfo{})) + uint32(unsafe.Sizeof(TraceProviderInfo{}))*0 // As there are no providers
+		if *pBufferSize < requiredSize {
+			// Set the size required and return the error
+			*pBufferSize = requiredSize
+			return ERROR_INSUFFICIENT_BUFFER
+		}
+
+		// Empty list of providers
+		*pBuffer = ProviderEnumerationInfo{
+			NumberOfProviders:      0,
+			TraceProviderInfoArray: [ANYSIZE_ARRAY]TraceProviderInfo{},
+		}
+		return nil
+	}
+
+	guid, err := GUIDFromProviderName(mockProviderName)
+	assert.EqualError(t, err, "no providers found")
+	assert.Equal(t, GUID{}, guid, "GUID should be empty when the provider is not found")
+}
+
+func TestGUIDFromProviderName_GUIDNotFound(t *testing.T) {
+	// Backup and defer restoration of the original function
+	originalFunc := EnumerateProvidersFunc
+	t.Cleanup(func() {
+		EnumerateProvidersFunc = originalFunc
```

Comment:
Mocked functions that are defined in the `Session` struct don't need a cleanup since every time a new session is created (in `NewSession()`), real functions are assigned to these pointers.

Functions  `_TdhEnumerateProviders`, `setSessionGUID`, and `guidFromProviderName` work a bit differently since it doesn't make sense to add them to the `Session` struct. That means they need to be restored to their original values every time a test uses them.

Considering your suggestion, I could create a helper function that restores every function pointer but it should be called from any test anyway. With the current implementation, only functions that are mocked are restored keeping each test independent from the others.